torch_geometric/utils/scatter.py

```diff
@@ -1,66 +1,91 @@
-import warnings
 from typing import Optional
 
 import torch
-import torch_scatter
 from torch import Tensor
 
-import torch_geometric
-
 major, minor, _ = torch.__version__.split('.', maxsplit=2)
 major, minor = int(major), int(minor)
 has_pytorch112 = major > 1 or (major == 1 and minor >= 12)
 
+if has_pytorch112:  # pragma: no cover
 
-class ScatterImpl(torch.nn.Module):
-    def forward(self, src: Tensor, index: Tensor, dim: int = -1,
+    class ScatterHelpers:
+        @staticmethod
+        def broadcast(src: Tensor, other: Tensor, dim: int) -> Tensor:
+            dim = other.dim() + dim if dim < 0 else dim
+            if src.dim() == 1:
+                for _ in range(0, dim):
+                    src = src.unsqueeze(0)
+            for _ in range(src.dim(), other.dim()):
+                src = src.unsqueeze(-1)
+            src = src.expand(other.size())
+            return src
+
+        @staticmethod
+        def generate_out(src: Tensor, index: Tensor, dim: int,
+                         dim_size: Optional[int]) -> Tensor:
+            size = list(src.size())
+            if dim_size is not None:
+                size[dim] = dim_size
+            elif index.numel() > 0:
+                size[dim] = int(index.max()) + 1
+            else:
+                size[dim] = 0
+            return src.new_zeros(size)
+
+        @staticmethod
+        def scatter_mean(src: Tensor, index: Tensor, dim: int, out: Tensor,
+                         dim_size: Optional[int]) -> Tensor:
+            out.scatter_add_(dim, index, src)
+
+            index_dim = dim
```

Comment:
Can we add some comment here what's going on with `index_dim`?

Reply:
Done.